miden-crypto/src/dsa/ecdsa_secp256k1/mod.rs

```diff
@@ -0,0 +1,574 @@
+            .sign_prehash_recoverable(&message_digest)
+            .expect("failed to generate signature");
+
+        let (r, s) = signature_inner.split_scalars();
+        let signature_data = SignatureData {
+            r: r.to_bytes().into(),
+            s: s.to_bytes().into(),
+            v: recovery_id.into(),
+        };
+
+        match hasher {
+            EcdsaHasher::Sha256 => Signature::Sha256(signature_data),
+            EcdsaHasher::Keccak => Signature::Keccak(signature_data),
+        }
+    }
+
+    /// Computes a Diffie-Hellman shared secret from this secret key and the ephemeral public key
+    /// generated by the other party.
+    pub fn get_shared_secret(&self, pk_e: EphemeralPublicKey) -> SharedSecret {
+        let shared_secret_inner = diffie_hellman(self.inner.as_nonzero_scalar(), pk_e.as_affine());
+
+        SharedSecret::new(shared_secret_inner)
+    }
+}
+
+// PUBLIC KEY
+// ================================================================================================
+
+/// Public key for ECDSA signature verification over secp256k1 curve.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PublicKey {
+    inner: VerifyingKey,
+}
+
+impl PublicKey {
+    /// Returns a commitment to the public key using the RPO256 hash function.
+    pub fn to_commitment(&self) -> Word {
+        <Self as SequentialCommit>::to_commitment(self)
+    }
+
+    /// Verifies a signature against this public key and message.
+    pub fn verify(&self, message: Word, signature: &Signature) -> bool {
+        let message_bytes: [u8; 32] = message.into();
+        let message_digest = signature.hasher().hash(&message_bytes);
+        let signature_inner = k256::ecdsa::Signature::from_scalars(*signature.r(), *signature.s());
+
+        match signature_inner {
+            Ok(signature) => self.inner.verify_prehash(&message_digest, &signature).is_ok(),
+            Err(_) => false,
+        }
+    }
+
+    /// Recovers from the signature the public key associated to the secret key used to sign the
+    /// message.
+    pub fn recover_from(message: Word, signature: &Signature) -> Self {
+        let message_bytes: [u8; 32] = message.into();
+        let message_digest = signature.hasher().hash(&message_bytes);
+        let signature_data = k256::ecdsa::Signature::from_scalars(*signature.r(), *signature.s())
+            .expect("could not build the signature from the scalars");
+
+        let verifying_key = k256::ecdsa::VerifyingKey::recover_from_prehash(
+            &message_digest,
+            &signature_data,
+            RecoveryId::from_byte(signature.v()).expect("invalid recovery id"),
+        )
+        .expect("failed to recover the public key from the message and signature");
+
+        Self { inner: verifying_key }
+    }
+}
+
+impl SequentialCommit for PublicKey {
+    type Commitment = Word;
+
+    fn to_elements(&self) -> Vec<Felt> {
+        self.to_bytes().chunks(7).map(Felt::from_bytes_with_padding).collect()
```

Comment:
Nit: I had to stare at this line for a minute, a code comment along the lines of "5 Felts representing 7 bytes each for 33 bytes of source data" would have made this easier.

Reply:
Agreed, I have added a comment about "why 7" which is the main obscure decision here